### PR TITLE
[ALLUXIO-2454] Hive permission issue fix  

### PR DESCRIPTION
**Pull request by peisun1115**
https://alluxio.atlassian.net/browse/ALLUXIO-2454

### CODE REVIEW

**Comment by peisun1115** [comment]
@calvinjia @ifcharming 

There are still lots of tests to fix. But can you take a look to see whether this kind of change is OK to you guys? I've changed a lot on the FileSystemContext code.  if no objections, i will go ahead and fix all the tests and also check whether it fix the problem.

**Comment by peisun1115** [comment]
Jenkins, test this please

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 1]
```diff
@@ -479,16 +480,45 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
-      return;
+    boolean sameSubject = false;
+    if (subject == null) {
+      sameSubject = newSubject == null;
+    } else {
+      sameSubject = subject.equals(newSubject);
     }
+    return sameHost && samePort && sameSubject;
+    /*
     throw new IOException(ExceptionMessage.DIFFERENT_MASTER_ADDRESS
```
Comment (ifcharming) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: do you intend to comment out this `IOException` ?

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 2]
```diff
@@ -479,16 +480,45 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
         .getMessage(mUri.getHost() + ":" + mUri.getPort(), masterAddress));
+        */
+  }
+
+  /**
+   * @return the hadoop subject if exists
+   */
+  private Subject getHadoopSubject() {
+    try {
+      UserGroupInformation ugi = UserGroupInformation.getCurrentUser();
```
Comment (ifcharming) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: is it possible for the current user to be different from the real user? when superuser proxy is used in UGI. Shall we use `UserGroupInformation.getRealUser()` in some cases? 

https://hadoop.apache.org/docs/r2.7.2/hadoop-project-dist/hadoop-common/Superusers.html

File: core/client/src/test/java/alluxio/hadoop/AbstractFileSystemTest.java [hunk 3]
```diff
@@ -288,19 +288,23 @@ public void getStatus() throws Exception {
...
+    PowerMockito.when(FileSystemContext.create(Mockito.any(Subject.class)))
+        .thenReturn(mMockFileSystemContext);
     mMockFileSystemMasterClient = Mockito.mock(FileSystemMasterClient.class);
     Mockito.when(mMockFileSystemContext.acquireMasterClient())
         .thenReturn(mMockFileSystemMasterClient);
+    Mockito.doNothing().when(mMockFileSystemMasterClient).connect();
   }
 
-  private void mockUserGroupInformation() throws IOException {
+  private void mockUserGroupInformation(String username) throws IOException {
```
Comment (ifcharming) [inline core/client/src/test/java/alluxio/hadoop/AbstractFileSystemTest.java]: not sure whether it's easy to create a unit test to simulate a UGI with proxy user, which might be helpful to verify the comments regarding `getRealUser` above

Comment (peisun1115) [inline core/client/src/test/java/alluxio/hadoop/AbstractFileSystemTest.java]: i tried it and have not made it working.  Will spend more time on it but not a high priority for now

File: core/client/src/main/java/alluxio/client/block/BufferedBlockOutStream.java [hunk 4]
```diff
@@ -60,14 +61,14 @@
    *
    * @param blockId the id of the block
    * @param blockSize the size of the block
-   * @param blockStoreContext the block store context
+   * @param context the block store context
```
Comment (ifcharming) [inline core/client/src/main/java/alluxio/client/block/BufferedBlockOutStream.java]: update comment here as well?

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java [hunk 5]
```diff
@@ -51,23 +52,23 @@
    * @param blockId the block id
    * @param blockSize the block size
    * @param workerNetAddress the address of the local worker
-   * @param blockStoreContext the block store context
+   * @param context the block store context
```
Comment (ifcharming) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: ditto

File: core/client/src/main/java/alluxio/client/block/RemoteBlockInStream.java [hunk 6]
```diff
@@ -46,7 +47,7 @@
   /** Client to communicate with the remote worker. */
   private final BlockWorkerClient mBlockWorkerClient;
   /** The block store context which provides block worker clients. */
```
Comment (ifcharming) [inline core/client/src/main/java/alluxio/client/block/RemoteBlockInStream.java]: ditto

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 2 above)
Comment (ifcharming) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: SG

File: core/client/src/test/java/alluxio/hadoop/AbstractFileSystemTest.java (hunk 3 above)
Comment (ifcharming) [inline core/client/src/test/java/alluxio/hadoop/AbstractFileSystemTest.java]: no problem

File: core/client/src/main/java/alluxio/client/block/BufferedBlockOutStream.java (hunk 4 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/BufferedBlockOutStream.java]: done

File: core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java (hunk 5 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/LocalBlockOutStream.java]: done

File: core/client/src/main/java/alluxio/client/block/RemoteBlockInStream.java (hunk 6 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/RemoteBlockInStream.java]: done

**Comment by peisun1115** [comment]
@ifcharming I have addressed your comments.

File: core/client/src/main/java/alluxio/client/RemoteBlockReader.java [hunk 7]
```diff
@@ -34,8 +35,8 @@ private Factory() {} // prevent instantiation
      *
      * @return a new instance of {@link RemoteBlockReader}
      */
-    public static RemoteBlockReader create() {
-      return new NettyRemoteBlockReader();
+    public static RemoteBlockReader create(FileSystemContext context) {
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/RemoteBlockReader.java]: It seems strange that the block reader needs a file system context.

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 8]
```diff
@@ -42,27 +58,143 @@
  * so, because thread A holds the lock on {@link FileSystemContext}.
  */
 @ThreadSafe
-public enum FileSystemContext {
-  INSTANCE;
+public class FileSystemContext {
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: Is this still a `FileSystem` context or more of a `Client` context?

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 9]
```diff
@@ -42,27 +58,143 @@
...
+  public static final FileSystemContext INSTANCE = new FileSystemContext(null);
+
+  static {
+    MetricsSystem.startSinks();
+    Metrics.initializeGauges();
+  }
 
-  private BlockStoreContext mBlockStoreContext;
+  // Master contexts.
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: `Client pools`?

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 10]
```diff
@@ -42,27 +58,143 @@
...
+  /**
+   * Creates a file system context with a subject.
+   *
+   * @param subject the parent subject, set to null if not present
    */
-  FileSystemContext() {
-    mFileSystemMasterClientPool = new FileSystemMasterClientPool(ClientContext.getMasterAddress());
-    mBlockStoreContext = BlockStoreContext.get();
-    mAlluxioBlockStore = new AlluxioBlockStore(mBlockStoreContext);
+  public FileSystemContext(Subject subject) {
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: Should this be private?

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 11]
```diff
@@ -42,27 +58,143 @@
...
+    int masterPort = Configuration.getInt(PropertyKey.MASTER_RPC_PORT);
+    mMasterAddress = new InetSocketAddress(masterHostName, masterPort);
+
+    mFileSystemMasterClientPool = new FileSystemMasterClientPool(mParentSubject, mMasterAddress);
+    mBlockMasterClientPool = new BlockMasterClientPool(mParentSubject, mMasterAddress);
+  }
+
+  /**
+   * Closes all the resources asscoated with the context. Make sure all the resources are released
+   * back to this context before calling this close. Usually, you don't need to call this.
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: Could you clarify when this should be called? Also, is this threadsafe?

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 12]
```diff
@@ -75,6 +207,15 @@ public FileSystemMasterClient acquireMasterClient() {
   }
 
   /**
+   * Releases a block master client into the block master client pool.
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: `file system`?

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 13]
```diff
@@ -90,60 +231,165 @@ public void close() {
...
+
+    if (!mFileSystemWorkerClientHeartbeatPools.containsKey(rpcAddress)) {
+      FileSystemWorkerThriftClientPool pool =
+          new FileSystemWorkerThriftClientPool(mParentSubject, rpcAddress,
+              Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_SIZE_MAX),
+              Configuration.getLong(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_GC_THRESHOLD_MS));
+      if (mFileSystemWorkerClientHeartbeatPools.putIfAbsent(rpcAddress, pool) != null) {
+        pool.close();
+      }
+    }
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: This kind of logic is repeated for all the different pools, is it possible to abstract it to a method?

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 14]
```diff
@@ -90,60 +231,165 @@ public void close() {
...
-   * @return the block store context
+   * Releases a netty channel to the channel pools.
+   *
+   * @param address the network address of the channel
+   * @param channel the channel to release
    */
-  public BlockStoreContext getBlockStoreContext() {
-    return mBlockStoreContext;
+  public void releaseNettyChannel(InetSocketAddress address, Channel channel) {
+    Preconditions.checkArgument(mNettyChannelPools.containsKey(address));
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: If a channel is released after the pool is closed, would we hit this condition, and if so, should we handle it gracefully?

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 13 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: I have added a TODO. I have thought about this for a while. No clean approach so far. I can address it later.

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 10 above)
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: You should be able to mock a final class in tests with Powermock.

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 13 above)
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: Sure

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 15]
```diff
@@ -437,58 +433,119 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
+   *
+   * @param uri the uri
+   * @param conf the hadoop conf
+   * @throws IOException if it fails to initialize
+   */
+  void initializeInternal(URI uri, org.apache.hadoop.conf.Configuration conf) throws IOException {
+    // Load Alluxio configuration if any and merge to the one in Alluxio file system. These
+    // modifications to ClientContext are global, affecting all Alluxio clients in this JVM.
+    // We assume here that all clients use the same configuration.
+    ConfUtils.mergeHadoopConfiguration(conf);
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: We won't merge the hadoop conf if `masterAddIsSameAsDefault` (since `initializeInternal` will not be called), is that ok?

Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: good point. thanks for catching this. I've reverted this change. It is at least called once now.

**Comment by peisun1115** [comment]
@calvinjia i've addressed your comments. PTAL. Thank you.

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 16]
```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
-      checkMasterAddress();
+      if (!masterAddIsSameAsDefault) {
+        throw new IOException(ExceptionMessage.DIFFERENT_MASTER_ADDRESS
+            .getMessage(mUri.getHost() + ":" + mUri.getPort(),
+                FileSystemContext.INSTANCE.getMasterAddress()));
+      }
+      setFileSystemAndContext();
       return;
     }
     synchronized (INIT_LOCK) {
```
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: is it still meaningful to check static bool `INIT_LOCK` here, since we don't have filesystem handler and context static any more.

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 17]
```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
+    Configuration.set(PropertyKey.MASTER_RPC_PORT, Integer.toString(uri.getPort()));
+    Configuration.set(PropertyKey.ZOOKEEPER_ENABLED, Boolean.toString(isZookeeperMode()));
+
+    // These must be reset to pick up the change to the master address.
+    // TODO(andrew): We should reset key value system in this situation - see ALLUXIO-1706.
+    LineageContext.INSTANCE.reset();
+    FileSystemContext.INSTANCE.reset();
+
+    // Try to connect to master, if it fails, the provided uri is invalid.
+    FileSystemMasterClient client = FileSystemContext.INSTANCE.acquireMasterClient();
```
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]:
```java
try (FileSystemContext.INSTANCE.acquireMasterClientResource()) {
}
```

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 18]
```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
     boolean samePort = masterAddress.getPort() == mUri.getPort();
     if (sameHost && samePort) {
-      return;
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @return the hadoop subject if exists
```
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: add "null otherwise". let's be specific if we could return null

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 19]
```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
+    LineageContext.INSTANCE.reset();
+    FileSystemContext.INSTANCE.reset();
+
+    // Try to connect to master, if it fails, the provided uri is invalid.
+    FileSystemMasterClient client = FileSystemContext.INSTANCE.acquireMasterClient();
+    try {
+      client.connect();
+      // Connected, initialize.
+    } catch (ConnectionFailedException | IOException e) {
+      LOG.error("Failed to connect to the provided master address {}: {}.",
```
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: unnecessary LOG before throw. LOG and throw is considered anti-pattern

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 20]
```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
+   * @param conf the hadoop conf
+   * @throws IOException if it fails to initialize
+   */
+  void initializeInternal(URI uri, org.apache.hadoop.conf.Configuration conf) throws IOException {
+    // Load Alluxio configuration if any and merge to the one in Alluxio file system. These
+    // modifications to ClientContext are global, affecting all Alluxio clients in this JVM.
+    // We assume here that all clients use the same configuration.
+    ConfUtils.mergeHadoopConfiguration(conf);
+    Configuration.set(PropertyKey.MASTER_HOSTNAME, uri.getHost());
+    Configuration.set(PropertyKey.MASTER_RPC_PORT, Integer.toString(uri.getPort()));
```
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: unnecessary `Integer.toString`

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 21]
```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
+    Configuration.set(PropertyKey.ZOOKEEPER_ENABLED, Boolean.toString(isZookeeperMode()));
```
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: same here, unnecessary `Boolean.toString`

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 22]
```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
+      throw new IOException(e);
+    } finally {
+      FileSystemContext.INSTANCE.releaseMasterClient(client);
+    }
+  }
+
+  /**
+   * Sets the file system and context.
+   */
+  void setFileSystemAndContext() {
```
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: why package private? It is not used by unit test

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 23]
```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
...
+      LOG.error("Failed to connect to the provided master address {}: {}.",
+          uri.toString(), e.toString());
+      throw new IOException(e);
+    } finally {
+      FileSystemContext.INSTANCE.releaseMasterClient(client);
+    }
+  }
+
+  /**
+   * Sets the file system and context.
```
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: NIT: how about calling this method `updateFileSystemAndContext`? Also can you describe what the method does with a bit more details?

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 24]
```diff
@@ -437,58 +433,117 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
     mStatistics = statistics;
     mUri = URI.create(mAlluxioHeader);
 
+    boolean masterAddIsSameAsDefault = checkMasterAddress();
```
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: update the javadoc

File: core/client/src/main/java/alluxio/client/UnderFileSystemFileReader.java [hunk 25]
```diff
@@ -37,16 +34,11 @@ private Factory() {} // prevent instantiation
...
      * Factory for {@link UnderFileSystemFileReader}.
      *
+     * @param context the file system context
      * @return a new instance of {@link UnderFileSystemFileReader}
      */
-    public static UnderFileSystemFileReader create() {
-      try {
-        return CommonUtils.createNewClassInstance(
-            Configuration.<UnderFileSystemFileReader>getClass(
-                PropertyKey.USER_UFS_FILE_READER_CLASS), null, null);
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/UnderFileSystemFileReader.java]: done

File: core/client/src/main/java/alluxio/client/UnderFileSystemFileWriter.java [hunk 26]
```diff
@@ -38,14 +35,8 @@ private Factory() {} // prevent instantiation
      *
      * @return a new instance of {@link UnderFileSystemFileWriter}
      */
-    public static UnderFileSystemFileWriter create() {
-      try {
-        return CommonUtils.createNewClassInstance(
-            Configuration.<UnderFileSystemFileWriter>getClass(
-                PropertyKey.USER_UFS_FILE_WRITER_CLASS), null, null);
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/UnderFileSystemFileWriter.java]: done

File: core/client/src/main/java/alluxio/client/file/FileOutStream.java [hunk 27]
```diff
@@ -92,10 +94,12 @@
    *
    * @param path the file path
    * @param options the client options
+   * @param context the file system context
    * @throws IOException if an I/O error occurs
    */
-  public FileOutStream(AlluxioURI path, OutStreamOptions options) throws IOException {
-    this(path, options, FileSystemContext.INSTANCE, UnderFileSystemFileOutStream.Factory.get());
+  public FileOutStream(AlluxioURI path, OutStreamOptions options, FileSystemContext context)
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileOutStream.java]: done

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 28]
```diff
@@ -42,27 +58,146 @@
  * so, because thread A holds the lock on {@link FileSystemContext}.
  */
 @ThreadSafe
-public enum FileSystemContext {
-  INSTANCE;
+public class FileSystemContext {
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: done

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 29]
```diff
@@ -42,27 +58,146 @@
...
+  public static final FileSystemContext INSTANCE = new FileSystemContext(null);
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: i can send another PR to rename it. If you don't feel very strong about this, I will just use INSTANCE for now. Too many places are affected.

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 30]
```diff
@@ -42,27 +58,146 @@
...
+   * lazily.
+   */
+  @GuardedBy("this")
+  private Boolean mHasLocalWorker;
+
+  /** The parent user associated with the {@link FileSystemContext}. */
+  private final Subject mParentSubject;
 
   /**
    * Creates a new file stream context.
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: done

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 31]
```diff
@@ -42,27 +58,146 @@
...
+  public static FileSystemContext create(Subject subject) {
+    return new FileSystemContext(subject);
+  }
+
+  /**
+   * Creates a file system context with a subject.
+   *
+   * @param subject the parent subject, set to null if not present
+   */
+  public FileSystemContext(Subject subject) {
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: done

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 32]
```diff
@@ -42,27 +58,146 @@
...
+    mParentSubject = subject;
+    init();
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: done

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 33]
```diff
@@ -42,27 +58,146 @@
...
+    mBlockMasterClientPool = new BlockMasterClientPool(mParentSubject, mMasterAddress);
+  }
+
+  /**
+   * Closes all the resources associated with the context. Make sure all the resources are released
+   * back to this context before calling this close. After closing the context, all the resources
+   * that acquired from this context might fail. Only call this when you are done with using
+   * the {@link FileSystem} associated with this {@link FileSystemContext}.
+   */
+  public void close() {
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: done

File: core/client/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 34]
```diff
@@ -90,60 +234,166 @@ public void close() {
...
+
+  /**
+   * Creates a client for a worker with the given address.
+   *
+   * @param address the address of the worker to get a client to
+   * @return a {@link BlockWorkerClient} connected to the worker with the given worker RPC address
+   * @throws IOException if it fails to create a client for a given hostname (e.g. no Alluxio
+   *         worker is available for the given worker RPC address)
+   */
+  public BlockWorkerClient createBlockWorkerClient(WorkerNetAddress address) throws IOException {
```
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/FileSystemContext.java]: done

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 24 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done. I feel it is not accurate. but i have added a comment as discussed.

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 20 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 21 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 17 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: I will keep it as it is for ease of mocking. let me know if you have any concern.

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 19 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: I think it is fine to log here because we don't control the callers of initialize(). And we are not sure wheher they will log proper information. This is useful to debug for Alluxio.

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 22 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 23 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 18 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done

**Comment by peisun1115** [comment]
@apc999 I've addressed your comments. PTAL? thanks

**Comment by peisun1115** [comment]
Jenkins, test this please

File: core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java [hunk 35]
```diff
@@ -276,63 +281,7 @@ public void sessionHeartbeat() throws IOException, InterruptedException {
...
-   *         there is enough capacity
-   * @throws InterruptedException if it is interrupted
-   */
-  private FileSystemWorkerClientService.Client acquireInternal(
-      ConcurrentHashMapV8<InetSocketAddress, FileSystemWorkerThriftClientPool> pools)
-      throws IOException, InterruptedException {
-    if (!pools.containsKey(mWorkerRpcServerAddress)) {
-      FileSystemWorkerThriftClientPool pool =
-          new FileSystemWorkerThriftClientPool(mWorkerRpcServerAddress,
-              Configuration.getInt(PropertyKey.USER_FILE_WORKER_CLIENT_POOL_SIZE_MAX),
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java]: ack

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 19 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: SG

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 16 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: yes

File: core/client/src/main/java/alluxio/client/file/BaseFileSystem.java [hunk 36]
```diff
@@ -325,4 +325,11 @@ public void unmount(AlluxioURI path, UnmountOptions options)
       mFileSystemContext.releaseMasterClient(masterClient);
     }
   }
+
+  @Override
+  public void close() {
+    if (mFileSystemContext != FileSystemContext.INSTANCE) {
+      mFileSystemContext.close();
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/file/BaseFileSystem.java]: technically, it is possible to use one context to create multiple file system instances, right? In this case, will we have trouble if one file system instance is closed, causing the shared context also closed?

Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/file/BaseFileSystem.java]: good point. I've removed FileSystem#Close and leave the user to explicitly manage the lifetime of the context. 

So in AbstractFileSystem, I am closing the context there.

**Comment by peisun1115** [comment]
@apc999 I've addressed your comment. PTAL? Thanks

File: core/common/src/main/java/alluxio/security/authentication/TransportProvider.java [hunk 37]
```diff
@@ -77,6 +78,18 @@ public static TransportProvider create() {
   TTransport getClientTransport(InetSocketAddress serverAddress) throws IOException;
 
   /**
+   * Similar as {@link TransportProvider#getClientTransport(InetSocketAddress)} but it also
+   * specifies the {@link Subject} explicitly.
+   *
+   * @param subject the subject
```
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authentication/TransportProvider.java]: is it ok for this parameter to be `null`?

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 17 above)
Comment (apc999) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: I don't understand. I am proposing to use try-with-resource for this piece of code to replace try-catch-finally.
Is this related to mocking?

Comment (peisun1115) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: I think these two are equivalent. I prefer not to use "try-with-resource" because I need to mock the return value of "FileSystemContext.INSTANCE.acquireMasterClientResource()". 
check AbstractFileSystemTest#mockFileSystemContextAndMasterClient.

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 38]
```diff
@@ -171,7 +176,7 @@ public synchronized void connect() throws IOException, ConnectionFailedException
           RuntimeConstants.VERSION, getServiceName(), mMode, mAddress);
 
       TProtocol binaryProtocol =
-          new TBinaryProtocol(mTransportProvider.getClientTransport(mAddress));
+          new TBinaryProtocol(mTransportProvider.getClientTransport(mParentSubject, mAddress));
```
Comment (peisun1115) [inline core/common/src/main/java/alluxio/AbstractClient.java]: yes

File: core/common/src/main/java/alluxio/security/authentication/PlainSaslTransportProvider.java [hunk 39]
```diff
@@ -55,6 +58,24 @@ public TTransport getClientTransport(InetSocketAddress serverAddress) throws IOE
...
+    String password = "noPassword";
+
+    if (subject != null) {
+      Set<User> user = subject.getPrincipals(User.class);
+      if (user != null && !user.isEmpty()) {
+        username = user.iterator().next().getName();
+      }
+    }
+    if (username == null || username.isEmpty()) {
+      username = LoginUser.get().getName();
```
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authentication/PlainSaslTransportProvider.java]: why we still need `LoginUser`?

Comment (peisun1115) [inline core/common/src/main/java/alluxio/security/authentication/PlainSaslTransportProvider.java]: yes. subject is not guaranteed to exist
